java-checks/src/main/java/org/sonar/java/checks/HasNextCallingNextCheck.java

```diff
@@ -54,7 +54,7 @@ public void visitNode(Tree tree) {
   }
 
   @Override
-  public void scanFile(JavaFileScannerContext context) {
+  public void leaveFile(JavaFileScannerContext context) {
```

Comment:
Cleaning the state in `leaveFile` is not guaranty, to ensure we don't face FP or FN because of this, we also need to clean the sate in `setContext`:
```java
  @Override
  public void setContext(JavaFileScannerContext context) {
    super.setContext(context);
    hasNextBodyVisitor.clearHasNextOwner();
  }
```

Reply:
good point